{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1085a466-3977-4cd5-8fac-5092f0b7e4c7",
   "metadata": {},
   "source": [
    "Ambient RNA detection\n",
    "----\n",
    "\n",
    "Perform ambient RNA detection with SoupX\n",
    "\n",
    "Environment: `preprocessing.yaml`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76790770-a79b-4584-9a12-87204aca4812",
   "metadata": {},
   "source": [
    "## Dependecy notebooks\n",
    "\n",
    "Input: \n",
    "\n",
    "Outpu: 36.372 × 34859\n",
    "\n",
    "- `02_quality_control.ipynb`\n",
    "\n",
    "Loads the qc filtered matrix. \n",
    "\n",
    "Returns: `adata_ambient_rna_corrected_matrix.h5ad` with ambient RNA genes removed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09fc633c-56c4-4d1f-953a-446c99870bd3",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "90b7f538-2f4d-4ceb-ad28-3f86cf7b7b09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "preprocessing\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "print(os.environ.get('CONDA_DEFAULT_ENV'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "60e798d0-84dc-4b15-8559-d7bd633d47d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "90ed9ec5-1151-495d-b41d-2a582761f231",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(SoupX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a07e7b3c-060a-40da-bce1-7eec0c1a6abc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "# import logging\n",
    "\n",
    "# import anndata2ri\n",
    "# import rpy2.rinterface_lib.callbacks as rcb\n",
    "# import rpy2.robjects as ro\n",
    "\n",
    "import scanpy as sc\n",
    "# import anndata as ad\n",
    "import numpy as np\n",
    "\n",
    "from pathlib import Path\n",
    "from rpy2.robjects import numpy2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "from scipy.sparse import csc_matrix\n",
    "\n",
    "# rcb.logger.setLevel(logging.ERROR)\n",
    "# ro.pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "\t\n",
    "import logging\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks as rcb\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "rcb.logger.setLevel(logging.ERROR)\n",
    "pandas2ri.activate()\n",
    "\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f540344f-3c16-4d38-bcc7-03fb860acf58",
   "metadata": {},
   "source": [
    "## Set up paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "951b5756-c085-4d4e-a80b-905c39b529bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"../\")  # this depends on the notebook depth and must be adapted per notebook\n",
    "\n",
    "from paths import DATA_DIR, RESULTS_DIR, FIG_DIR\n",
    "from utils.utils import SAMPLES, VERSION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a283f809-71c9-4982-b77a-24efc0cdd8c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 36372 × 78932\n",
       "    obs: 'sample', 'condition', 'area', 'age', 'donor', 'sequencing batch', 'sequencing id', 'sex', 'APOE genotype', 'brain area number', 'running number', 'donor_v2', 'donor_v2_area', 'n_counts', 'log_counts', 'n_genes', 'tech_sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_nuclear', 'log1p_total_counts_nuclear', 'pct_counts_nuclear', 'braak_stage', 'area_abbrev', 'condition_area', 'brainbank', 'brain_area_latin'\n",
       "    var: 'ensemble_id', 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'mt', 'ribo', 'hb', 'nuclear', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    uns: 'area_colors', 'brain area latin_colors', 'cell_type_sea_colors', 'condition_colors', 'leiden', 'leiden_colors', 'neighbors', 'pca', 'sample_colors', 'tech_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'X_log', 'ambiguous', 'log1p_norm', 'matrix', 'raw', 'spliced', 'unspliced'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(Path(DATA_DIR / f\"{VERSION}_adata_filtered_counts_matrix_final.h5ad\"))\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c886c43b-23e1-432a-828f-512de0efad03",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pp = adata.copy()\n",
    "sc.pp.normalize_total(adata_pp, target_sum=1e4)\n",
    "sc.pp.log1p(adata_pp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "328cc656-fdc6-4b6b-8c55-318186d10ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata_pp)\n",
    "sc.pp.neighbors(adata_pp)\n",
    "sc.tl.leiden(\n",
    "    adata_pp, key_added=\"soupx_groups\", flavor=\"igraph\", n_iterations=2, directed=False\n",
    ")\n",
    "# Preprocess variables for SoupX\n",
    "adata.obs[\"soupx_groups\"] = adata_pp.obs[\"soupx_groups\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "53a288f5-3776-4825-b457-017114b5561d",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bf471195-c041-4808-8a73-b61b8e89af01",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = adata.obs_names\n",
    "genes = adata.var_names\n",
    "data = adata.X.T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4157a63e-9918-47de-8ac4-32b5325f46dc",
   "metadata": {},
   "source": [
    "SoupX additionally requires the raw gene by cell matrix from `raw_feature_bc_matrix..h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eea8e9fb-2364-472a-a706-2eedc014b23f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 54957338 × 78932\n",
       "    obs: 'sample', 'condition', 'area', 'age', 'Braak stage', 'donor', 'sequencing batch', 'Brainbank source', 'sequencing id', 'sex', 'APOE genotype', 'brain area latin', 'brain area number', 'running number'\n",
       "    var: 'ensemble_id', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'\n",
       "    layers: 'raw'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw = sc.read(Path(DATA_DIR / f\"{VERSION}_adata_raw_feature_matrix.h5ad\"))\n",
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "409948f7-cfd3-4f5c-839f-fbfd42b38a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_raw = adata_raw.var_names\n",
    "cells_raw = adata_raw.obs_names\n",
    "\n",
    "data_tod = adata_raw.X.T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f2f403f-0f74-4770-a21e-4b648aee29f2",
   "metadata": {},
   "source": [
    "Check that the gene names are equavalent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "57e8f47f-819e-4a04-b343-fd806806615a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not (all(genes == genes_raw)):\n",
    "    # Genes in genes but not in genes_raw\n",
    "    diff_1 = set(genes) - set(genes_raw)\n",
    "    \n",
    "    # Genes in genes_raw but not in genes\n",
    "    diff_2 = set(genes_raw) - set(genes)\n",
    "    \n",
    "    print(\"In genes but not in genes_raw:\", len(diff_1), list(diff_1)[:10])\n",
    "    print(\"In genes_raw but not in genes:\", len(diff_2), list(diff_2)[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fd4ae7b0-cb36-4511-871a-e4b2e3e10d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes = np.array([g.replace('.', '-') for g in genes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6d02f811-affd-4295-98c3-6bcc483b822e",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_raw "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3fba0757-8b07-4328-889c-438b13a3f826",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_csc = data.tocsc()\n",
    "data_tod_csc = data_tod.tocsc()\n",
    "\n",
    "# Extract sparse components and cast to correct types\n",
    "x = data_csc.data.astype(np.float64)\n",
    "i = data_csc.indices.astype(np.int32)\n",
    "p = data_csc.indptr.astype(np.int32)\n",
    "dims = np.array(data_csc.shape, dtype=np.int32)\n",
    "\n",
    "x_tod = data_tod_csc.data.astype(np.float64)\n",
    "i_tod = data_tod_csc.indices.astype(np.int32)\n",
    "p_tod = data_tod_csc.indptr.astype(np.int32)\n",
    "dims_tod = np.array(data_tod_csc.shape, dtype=np.int32)\n",
    "\n",
    "with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):\n",
    "    ro.globalenv[\"x\"] = x\n",
    "    ro.globalenv[\"i\"] = i\n",
    "    ro.globalenv[\"p\"] = p\n",
    "    ro.globalenv[\"dims\"] = dims\n",
    "\n",
    "    ro.globalenv[\"x_tod\"] = x_tod\n",
    "    ro.globalenv[\"i_tod\"] = i_tod\n",
    "    ro.globalenv[\"p_tod\"] = p_tod\n",
    "    ro.globalenv[\"dims_tod\"] = dims_tod\n",
    "\n",
    "    ro.globalenv[\"genes\"] = np.array(genes)\n",
    "    ro.globalenv[\"genes_raw\"] = np.array(genes_raw)\n",
    "    ro.globalenv[\"cells\"] = np.array(cells)\n",
    "    ro.globalenv[\"cells_raw\"] = np.array(cells_raw)\n",
    "    ro.globalenv[\"soupx_groups\"] = adata.obs[\"soupx_groups\"].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "be0808a1-10a2-4db0-bd36-a1c0bc89d643",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o out \n",
    "\n",
    "library(Matrix)\n",
    "\n",
    "# Manually coerce types to avoid \"array\" class errors\n",
    "x <- as.numeric(x)\n",
    "i <- as.integer(i)\n",
    "p <- as.integer(p)\n",
    "dims <- as.integer(dims)\n",
    "\n",
    "x_tod <- as.numeric(x_tod)\n",
    "i_tod <- as.integer(i_tod)\n",
    "p_tod <- as.integer(p_tod)\n",
    "dims_tod <- as.integer(dims_tod)\n",
    "\n",
    "# Reconstruct sparse matrices\n",
    "data <- new(\"dgCMatrix\",\n",
    "            Dim = dims,\n",
    "            x = x,\n",
    "            i = i,\n",
    "            p = p)\n",
    "\n",
    "data_tod <- new(\"dgCMatrix\",\n",
    "                Dim = dims_tod,\n",
    "                x = x_tod,\n",
    "                i = i_tod,\n",
    "                p = p_tod)\n",
    "\n",
    "# Assign row and column names\n",
    "rownames(data) <- genes\n",
    "colnames(data) <- cells\n",
    "rownames(data_tod) <- genes_raw\n",
    "colnames(data_tod) <- cells_raw\n",
    "\n",
    "# SoupX pipeline\n",
    "sc = SoupChannel(data_tod, data, calcSoupProfile = FALSE)\n",
    "\n",
    "soupProf = data.frame(\n",
    "  row.names = rownames(data),\n",
    "  est = rowSums(data) / sum(data),\n",
    "  counts = rowSums(data)\n",
    ")\n",
    "sc = setSoupProfile(sc, soupProf)\n",
    "sc = setClusters(sc, soupx_groups)\n",
    "sc = autoEstCont(sc, doPlot = FALSE)\n",
    "out = adjustCounts(sc, roundToInt = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38b0d25c-c978-4fa3-a190-10f45c259847",
   "metadata": {},
   "source": [
    "Store the SoupX corrected counts. Store old counts in additional layer and overwrite the SoupX counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f0e05fd5-5823-4d21-b1dd-05ac976d4288",
   "metadata": {},
   "outputs": [],
   "source": [
    "with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):\n",
    "    out_py = ro.conversion.rpy2py(ro.globalenv[\"out\"])\n",
    "\n",
    "x = np.array(out_py.slots[\"x\"])\n",
    "i = np.array(out_py.slots[\"i\"])\n",
    "p = np.array(out_py.slots[\"p\"])\n",
    "shape = tuple(out_py.slots[\"Dim\"])\n",
    "\n",
    "out_matrix = csc_matrix((x, i, p), shape=shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ce305907-e6ab-46c0-9367-a808120b565f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "adata.layers[\"soupX_counts\"] = out_matrix.T\n",
    "adata.X = adata.layers[\"soupX_counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "787f6711-6eef-4b51-91b0-66bbd5ae1f2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of genes: 78932\n",
      "Number of genes after cell filter: 34859\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total number of genes: {adata.n_vars}\")\n",
    "\n",
    "# Min 20 cells - filters out 0 count genes\n",
    "sc.pp.filter_genes(adata, min_cells=20)\n",
    "print(f\"Number of genes after cell filter: {adata.n_vars}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edce971-71bb-457f-b715-b344986fb84f",
   "metadata": {},
   "source": [
    "## Save adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "38e118be-6659-4365-b2de-ada66d64e519",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 36372 × 34859\n",
       "    obs: 'sample', 'condition', 'area', 'age', 'donor', 'sequencing batch', 'sequencing id', 'sex', 'APOE genotype', 'brain area number', 'running number', 'donor_v2', 'donor_v2_area', 'n_counts', 'log_counts', 'n_genes', 'tech_sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_nuclear', 'log1p_total_counts_nuclear', 'pct_counts_nuclear', 'braak_stage', 'area_abbrev', 'condition_area', 'brainbank', 'brain_area_latin', 'soupx_groups'\n",
       "    var: 'ensemble_id', 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'mt', 'ribo', 'hb', 'nuclear', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'\n",
       "    uns: 'area_colors', 'brain area latin_colors', 'cell_type_sea_colors', 'condition_colors', 'leiden', 'leiden_colors', 'neighbors', 'pca', 'sample_colors', 'tech_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'X_log', 'ambiguous', 'log1p_norm', 'matrix', 'raw', 'spliced', 'unspliced', 'counts', 'soupX_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "03aa9176-c619-4c29-add4-e646ba57027a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(Path(DATA_DIR / f\"{VERSION}_adata_ambient_rna_corrected_matrix.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40ab6c4e-951b-455f-b2c2-e1eec6bda523",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
